ksmt-symfpu/src/main/kotlin/org/ksmt/symfpu/FpToBvTransformer.kt

```diff
@@ -0,0 +1,463 @@
+        transformExprAfterTransformed(expr, expr.lhs, expr.rhs) { l, r ->
+            if (l is UnpackedFp<*> && r is UnpackedFp<*>) {
+                val flags = mkAnd(l.isNaN eq r.isNaN, l.isInf eq r.isInf, l.isZero eq r.isZero)
+                if (l.packedBv is UnpackedFp.PackedFp.Exists && r.packedBv is UnpackedFp.PackedFp.Exists)
+                    flags and (l.packedBv eq r.packedBv)
+                else mkAnd(
+                    flags,
+                    l.sign eq r.sign,
+                    l.unbiasedExponent eq r.unbiasedExponent,
+                    l.normalizedSignificand eq r.normalizedSignificand,
+                )
+            } else {
+                l eq r
+            }
+        }
+    }
+
+    override fun <T : KSort> transform(expr: KIteExpr<T>): KExpr<T> = with(ctx) {
+        transformExprAfterTransformed(expr, expr.condition, expr.trueBranch, expr.falseBranch) { c, l, r ->
+            if (l is UnpackedFp<*> && r is UnpackedFp<*>) {
+                val lTyped: UnpackedFp<KFpSort> = l.cast()
+                iteOp(c, lTyped, r.cast()).cast()
+            } else {
+                mkIte(c, l, r).cast()
+            }
+        }
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpEqualExpr<Fp>): KExpr<KBoolSort> = with(ctx) {
+        transformHelper(expr, ::equal)
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpLessExpr<Fp>): KExpr<KBoolSort> = with(ctx) {
+        transformHelper(expr, ::less)
+    }
+
+    override fun <Fp : KFpSort> transform(expr: KFpMulExpr<Fp>): KExpr<Fp> = with(ctx) {
+        val args1: List<KExpr<Fp>> = expr.args.cast()
+        transformExprAfterTransformed(expr, args1) { args ->
+            val (left, right) = argsToTypedPair(args.drop(1))
+            multiply(left, right, args[0].cast())
+        }
+    }
+
+    override fun <A : KArraySortBase<R>, R : KSort> transform(expr: KArrayConst<A, R>): KExpr<A> = with(ctx) {
+        transformExprAfterTransformed(expr, expr.value) { value ->
+            println("transforming arrayConst")
+            val resSort = transformedArraySort(expr.cast())
+            mkArrayConst(resSort, packToBvIfUnpacked(value)).cast()
+        }
+    }
+
+
+    override fun <D : KSort, R : KSort> transform(expr: KArraySelect<D, R>): KExpr<R> {
+        return transformExprAfterTransformed(expr, expr.array, expr.index) { array, index ->
+            with(ctx) {
+                arraysTransform.arraySelectUnpacked(expr.sort, array.select(packToBvIfUnpacked(index).cast()))
+            }
+        }
+    }
+
+
+    override fun <D : KSort, D1 : KSort, R : KSort> transform(expr: KArray2Select<D, D1, R>): KExpr<R> {
+        return transformExprAfterTransformed(expr, expr.array, expr.index0, expr.index1) { array, index0, index1 ->
+            with(ctx) {
+                arraysTransform.arraySelectUnpacked(expr.sort,
+                    array.select(packToBvIfUnpacked(index0).cast(), packToBvIfUnpacked(index1).cast()))
```

Comment:
Can we avoid these `.cast` everywhere? Maybe, we can change type of `packToBvIfUnpacked` function?